James/vendia controls (#27)
vendia/examples: features/share/access-controls/user-access-controls/README.md

<p align="center">
  <a href="https://vendia.net/">
    <img src="https://raw.githubusercontent.com/vendia/examples/main/vendia-logo.png" alt="vendia logo" width="100px">
  </a>
</p>

# User Access Controls

> User access controls protect the user account management operations provided by Vendia Share.  This prevents unauthorized access to user management functions and enables features like self-service user management within an organization.

Reviewer: ```suggestion
You will use User Access Controls to prevent unauthorized access to user management functions and to enable features such as self-service user management within an organization.
```
```suggestion
User access controls protect the user account management operations provided by Vendia Share.  This prevents unauthorized access to user management functions and enables features like self-service user management within an organization.
```

Reviewer: Try to merge the first and second sentences. They both seem to imply the same thing.  suggested a change.

Author: I'm using the first few sentences of each section to describe the feature, not what the user will do.  The intent is to provide them a bit of context before diving in to the "you will..." portion.  Let me know if that flow works.

Reviewer: the new version of the definition looks great. I noticed it because at all other places you kept it simple. This is the only place where I have to read 3-4 times to fully comprehend it.

Authorization to specific user account management operations that act on user accounts is granted through a [Role Based Access Control (RBAC) mechanism](https://www.vendia.net/docs/share/rbac).  Before going further, it's important to define a few key RBAC terms, as they relate to user access controls.

> ## Key Terminology

Reviewer: I found this section very dense. I will make a few suggestions. Trying to balance between 'how precise' and 'how simple'.

Author: Great call.  I simplified the language in the User and Uni sections, both here and in the blog post.

> * **Action** - The right to call one of the user account management operations exposed by the Vendia Share control plane

Reviewer: ```suggestion
* **Action** - defines the privilege to call the user account management operation on a target resource.
```
```suggestion
* **Action** - The right to call one of the user account management operations exposed by the Vendia Share control plane
```

Author: Reworked in latest

> * **Resource** - The Vendia Share account or accounts against which a user account management Action can be performed

Reviewer: ```suggestion
* **Resource** - is a target Vendia Share account(s) against which an Action is performed.
```
```suggestion
* **Resource** - The Vendia Share account or accounts against which a user account management Action can be performed
```

Author: Reworked in latest

> * **Capability** - A mapping of Action to Resource, linking the right to call (an operation) with the right to manage (a Vendia Share account)

Reviewer: ```suggestion
* **Capability** - links Action to a Resource
```
```suggestion
* **Capability** - A mapping of Action to Resource, linking the right to call (an operation) with the right to manage (a Vendia Share account)
```

Author: Reworked in latest

> * **Role** - A named collection of Capabilities; used to determine if a user management operation, executed by invoking the Vendia Share control plane API, is permitted or denied.

Reviewer: ```suggestion
* **Role** - is a collection of Capabilities; used to determine if a user management operation is permitted or denied.
```
```suggestion
* **Role** - A named collection of Capabilities; used to determine if a user management operation, executed by invoking the Vendia Share control plane API, is permitted or denied.
```

Author: Reworked in latest

UserId:  aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee
Has role: default

        Permitting: UNI_INVITE
        On: NameResource(you@*.domain.com)

        Permitting: UNI_QUERY
        On: NameResource(you@*.domain.com)

        Permitting: UNI_RESET
        On: NameResource(you@*.domain.com)

        Permitting: UNI_DELETE
        On: NameResource(you@*.domain.com)

        Permitting: UNI_GET
        On: NameResource(you@*.domain.com)

        Permitting: UNI_JOIN
        On: UniResource(*.unis.vendia.net#*)

        Permitting: UNI_CREATE
        On: UniResource(*.unis.vendia.net#*)

        Permitting: UNI_EVOLVE_SCHEMA
        On: NameResource(you@*.domain.com)

        Permitting: UNI_MUTATE
        On: NameResource(you@*.domain.com)

        Permitting: UNI_DELETE_NODE
        On: NameResource(you@*.domain.com)

        Permitting: USER_INVITE
        On: NameResource(*@*.*.*)

```
</details>

You can inspect the `getRole.js` files behind the `npm` command above to see the code examples in detail, including the use of a Vendia-authored library called [aws-signed-fetch](https://www.npmjs.com/package/@vendia/aws-signed-fetch/v/0.0.5).

## Summary

In this example you:

> * Learned more about Vendia Share User Access Controls

Reviewer: Do you mean Key Terms and Concepts? Can we be more specific?

* Used the Share CLI to interact with some of Vendia Share's user management operations, which are protected by User Access Controls 

> * Used a Node.js GraphQL client to interact with those same user management operations, demonstrating the ability to fully automate user management with any GraphQL-capable tool you choose

Reviewer: just curious, "what is the reason for choosing js for scripting?" Usually, for admin tasks, python and shell are preferred. Are we observing that App Dev will be doing most of the Uni Management? Is that why?

Author: Just to mix in some more languages/platform examples throughout our `examples` repo.  We have an awful lot of python GraphQL client interactions already.  Node is very popular on the AppDev side so wanted to show some examples with that as well.
